fix(service-worker): improvements/fixes regarding ServiceWorker caches

gkalpak opened the pull request:
Several improvements and fixes regarding the ServiceWorker caches (naming, accessing, clean-up).
See individual commits for details.

Fixes #41728.

##
_NOTE: Theoretically, this could land on patch as well. But since it changes some of the SW caches names, let's be conservative and land in a minor release._

mary-poppins commented:
You can preview 2159d08 at https://pr42622-2159d08.ngbuilds.io/.

AndrewKushnir commented on packages/service-worker/worker/src/app-version.ts:
```diff
@@ -284,11 +284,14 @@ export class AppVersion implements UpdateSource {
-    await Promise.all(this.assetGroups.map(group => group.cleanup()));
-    await Promise.all(this.dataGroups.map(group => group.cleanup()));
+  async getCacheNames(): Promise<string[]> {
+    const [assetGroupCacheNames, dataGroupCacheNames] = await Promise.all([
+      Promise.all(this.assetGroups.map(group => group.getCacheNames())),
+      Promise.all(this.dataGroups.map(group => group.getCacheNames())),
+    ]);
+    return ([] as string[]).concat(...assetGroupCacheNames, ...dataGroupCacheNames);
```
Since we concat the data at the end, can we avoid nested `Promise.all` calls?
```
    return await Promise.all([
      ...this.assetGroups.map(group => group.getCacheNames()),
      ...this.dataGroups.map(group => group.getCacheNames()),
    ]);
```

AndrewKushnir reviewed (approved):
LGTM overall, I just left a few comments. The most important one is related to a possible typo in the `NamedCacheStorage.has` code.

AndrewKushnir commented:
@gkalpak FYI I was going through the commits one-by-one, so some of the comments are marked as "outdated" in GitHub UI, feel free to ignore them if they are not relevant and/or already resolved.

gkalpak commented on packages/service-worker/worker/src/assets.ts:
```diff
@@ -102,8 +103,8 @@ export abstract class AssetGroup {
-    await this.scope.caches.delete(`${this.prefix}:${this.config.name}:cache`);
-    await this.db.delete(`${this.prefix}:${this.config.name}:meta`);
+    await this.scope.caches.delete(`${this.cacheNamePrefix}:${this.config.name}:cache`);
+    await this.db.delete(`${this.cacheNamePrefix}:${this.config.name}:meta`);
```
Addressed in a subsequent commit.

gkalpak commented on packages/service-worker/worker/src/debug.ts:
```diff
@@ -9,6 +9,7 @@
 import {Adapter} from './adapter';
 import {Debuggable, DebugLogger} from './api';
 
+const SW_VERSION = '0.0.0-PLACEHOLDER';
```
The version is replaced when building the packages (by the [ng_package][1] and [pkg_npm][2] bazel rules). This is the standard mechanism we use to populate fields/variables with the current version all over the repo (for example, in `package.json` files and `VERSION` exports for all `@angular/*` packages, license banners, etc.).

I could add a comment, but I don't think it would help. It would be a drop in the ocean :smiley:
(For reference, the special `0.0.0-PLACEHOLDER` placeholder is used in ~100 files throughout the repo.)

Maybe we could add some docs on how version stamping works (I couldn't find anything from a quick search).

[1]: https://github.com/angular/angular/blob/a16f82b851c1cf1f5ba8739e574dfbc89a662666/tools/defaults.bzl#L178
[2]: https://github.com/angular/angular/blob/a16f82b851c1cf1f5ba8739e574dfbc89a662666/tools/defaults.bzl#L216

AndrewKushnir replied:
You are right! I think it's ok to keep as it as it's a standard thing that we do across the repo. Thanks for the clarification.

gkalpak commented on packages/service-worker/worker/src/db-cache.ts:
```diff
@@ -31,7 +31,7 @@ export class CacheDatabase implements Database {
     const allCacheNames = await this.scope.caches.keys();
     const dbCacheNames = allCacheNames.filter(name => name.startsWith(prefix));
-    return dbCacheNames;
+    return dbCacheNames.map(name => name.slice(prefix.length));
```
Good call :+1:

gkalpak commented on packages/service-worker/worker/src/data.ts:
```diff
@@ -249,11 +249,10 @@ export class DataGroup {
+    this.cache =
+        scope.caches.open(`${adapter.cacheNamePrefix}:${cacheNamePrefix}:${config.name}:cache`);
+    this.lruTable = this.db.open(`${cacheNamePrefix}:${config.name}:lru`, config.cacheQueryOptions);
+    this.ageTable = this.db.open(`${cacheNamePrefix}:${config.name}:age`, config.cacheQueryOptions);
```
After subsequent changes, these keys are now created in a single place.

gkalpak commented on packages/service-worker/worker/src/named-cache-storage.ts:
```diff
@@ -0,0 +1,43 @@
+ * A wrapper around `CacheStorage` to allow interacting with the caches more easily while
+ * name-spacing caches to avoid conflicts with other caches on the same domain.
+ */
+export class NamedCacheStorage<T extends CacheStorage> implements CacheStorage {
```
I see how my comment can be confusing as it puts more emphasis on the name-spacing part.

However, the main trait of this versus the native `CacheStorage` is that the returned `Cache` instances have a `name` property (which can be used to more easily delete them, for example).

Therefore, I have a slight preference for keeping the `Named` prefix. I will update the JSDoc comment to highlight that.
I don't feel strongly about it though, so happy to switch to `Namespaced` if you think it is preferable.

AndrewKushnir replied:
That makes sense, thanks @gkalpak 👍 Let's keep the `NamedCacheStorage` version that you have in the PR.

gkalpak commented on packages/service-worker/worker/src/service-worker.d.ts:
```diff
@@ -107,7 +107,9 @@ interface ExtendableMessageEvent extends ExtendableEvent {
-  caches: CacheStorage;
+  // Disallow accessing `CacheStorage APIs directly to ensure that all accesses go through a
+  // `NamedCacheStorage` instance (exposed by the `Adapter`).
+  //caches: CacheStorage;
```
I updated the comment to make it clear how we are disallowing direct access to `CacheStorage` APIs and removed the commented out code.

gkalpak commented on packages/service-worker/worker/src/named-cache-storage.ts:
```diff
@@ -0,0 +1,43 @@
+  }
+
+  has(cacheName: string): Promise<boolean> {
+    return this.original.delete(`${this.cacheNamePrefix}:${cacheName}`);
```
😱 
Fixed.

gkalpak commented on packages/service-worker/worker/src/named-cache-storage.ts:
```diff
@@ -0,0 +1,43 @@
+  async keys(): Promise<string[]> {
+    const prefix = `${this.cacheNamePrefix}:`;
+    const allCacheNames = await this.original.keys();
+    const ownCacheNames = allCacheNames.filter(name => name.startsWith(prefix));
+    return ownCacheNames.map(name => name.slice(prefix.length));
```
Yeah, this aspect is not worth optimizing here.

gkalpak commented on packages/service-worker/worker/src/named-cache-storage.ts:
```diff
@@ -0,0 +1,43 @@
+ * name-spacing caches to avoid conflicts with other caches on the same domain.
+ */
+export class NamedCacheStorage<T extends CacheStorage> implements CacheStorage {
+  constructor(readonly original: T, private cacheNamePrefix: string) {}
```
In some cases, we do need to access the original `CacheStorage` (both in tests and in actual code).
But at least we explicitly access it via the `NamedCacheStorage` instance (instead of directly via the `ServiceWorkerGlobalScope`), so it is clear that it is intentional.

gkalpak committed:
feat(service-worker): include ServiceWorker version in debug info

This commit includes the ServiceWorker version in the debug info shown
at `/ngsw/state` to make it easier to know what version of the
ServiceWorker script is controlling the page.

gkalpak committed:
fix(service-worker): improve ServiceWorker cache names

This commit improves the cache names generated by the ServiceWorker by
making them shorter and non-repetitive. In particular, the following
changes are made:

- Data-group cache names no longer include the `dynamic` infix, since it
  does not add any value.
  Before: `ngsw:<...>:data:dynamic:<...>`
  After:  `ngsw:<...>:data:<...>`

- `CacheDatabase` table names no longer include the `ngsw:<path>` prefix
  twice.
  Before: `ngsw:<path>:db:ngsw:<path>:<...>`
  After:  `ngsw:<path>:db:<...>`

NOTE 1:
This change will result in different cache names being generated for the
same app-versions with the new SericeWorker script. This means that some
of the previously cached data will need to be re-downloaded (because the
ServiceWorker will not be able to re-use the old caches), but that
should be transparent for the end user.
While possible, adding logic to allow the ServiceWorker to retrieve data
from the old caches is not worth the extra complecity and maintenance
cost.

NOTE 2:
Generating different cache names for some of the caches means that the
ServiceWorker will not be able to clean-up some of the old caches. This
will be taken care of in a subsequent commit that will rework the
clean-up logic to be more robust (covering changes such as this one and
other edgecases).

gkalpak committed:
fixup! fix(service-worker): use correct names when listing `CacheDatabase` tables

gkalpak committed:
refactor(service-worker): minor refactorings to improve readability/maintainability

This commit makes some minor refactorings to improve the code
readability and maintainability, including:
- Avoiding code duplication.
- Using more descriptive variable names.
- Using `async/await` instead of `Promise#then()`.
- Accessing variables directly instead of via `this` when possible.

gkalpak committed:
fix(service-worker): use correct names when listing `CacheDatabase` tables

`CacheDatabase` uses the un-prefixed table names to interact with
database tables. However, the `list()` method returns the raw, prefixed
table names (which are not useful, since they cannot be used to
open/delete a table).

This commit fixes this by removing the prefix from the cache names
returned by the `list()` method.

NOTE:
This method is currently not used anywhere, so this change does not
affect the ServiceWorker behavior.

gkalpak committed:
fixup! refactor(service-worker): simplify accessing `CacheStorage` throughout the ServiceWorker

gkalpak committed:
fixup! fix(service-worker): ensure obsolete caches are always cleaned up

gkalpak committed:
fix(service-worker): ensure caches are cleaned up when failing to load state

Previously, obsolete caches were only cleaned up when successfully
loading the stored state. When the state failed to be loaded, cleaning
up the caches would be skipped until the next SW initialization.

This commit changes this, ensuring that the caches are cleaned up
regardless if the stored state was loaded successfully or not.

gkalpak committed:
fix(service-worker): ensure obsolete caches are always cleaned up

Previously, the SW was only able to clean up caches for app-versions
found in the `Driver`'s `versions` map. If (for some reason) the
`Driver` failed to load a valid stored state (including app-versions)
and ended up with an [empty `versions` map][1], any obsolete versions
would remain in the cache storage. This case was rare but possible.

This commit makes the cache clean-up logic more robust by ensuring that
all app-version caches are removed unless they are currently used by the
SW to serve active clients (with the exception of the latest
app-version, which is always retained).

Fixes #41728

[1]: https://github.com/angular/angular/blob/9de65dbdceac3077881fbc49717f33d0f379e21d/packages/service-worker/worker/src/driver.ts#L515-L529

gkalpak committed:
refactor(service-worker): simplify accessing `CacheStorage` throughout the ServiceWorker

This commit simplifies/systemizes accessing the `CacheStorage` through a
wrapper, with the following benefits:
- Ensuring a consistent cache name prefix is used for all caches
  (without having to repeat the prefix in different places).
- Allowing referring to caches using their name without the common
  cache name prefix.
- Exposing the cache name on cache instances, which for example makes it
  easier to delete caches without having to keep track of the name used
  to create them.

gkalpak commented:
Thx for the review, @AndrewKushnir ✨
I have addressed most of the comments (in fixup commits) and replied to the rest. I have left a couple of review comments unresolved to get your input. PTAL

mary-poppins commented:
You can preview d0ddf1b at https://pr42622-d0ddf1b.ngbuilds.io/.

AndrewKushnir reviewed (approved):
👍

dylhunn commented:
N.B. This is currently blocked on the CI coming back.

mary-poppins commented:
You can preview 612ba53 at https://pr42622-612ba53.ngbuilds.io/.

gkalpak commented:
This should also have fixed #20785.
